software/contrib/quantizer.py

```diff
@@ -0,0 +1,537 @@
+        """
+        state = self.load_state_json()
+        
+        loaded_scale = state.get("scale", [True]*12)  # default to a chromatic scale
+        self.scale.notes = loaded_scale
+        
+        self.root = state.get("root", self.root)
+        self.octave = state.get("octave", self.octave)
+        self.intervals = state.get("intervals", self.intervals)
+        self.mode = state.get("mode", self.mode)
+    
+    def save(self):
+        """Save the current settings to persistent storage
+        """
+        state = {
+            "scale": self.scale.notes,
+            "root": self.root,
+            "octave": self.octave,
+            "intervals": self.intervals,
+            "mode": self.mode
+        }
+        self.save_state_json(state)
+        
+    @classmethod
+    def display_name(cls):
+        return "Quantizer"
+       
+    def quantize(self, analog_in):
+        """Take an analog signal and process it
+
+        Sets self.current_note and self.output_voltage
+        """
+        
+        (volts, semitone) = self.scale.quantize(analog_in)
+        
+        # apply our octave & transposition offsets
+        volts = volts + self.octave * VOLTS_PER_OCTAVE + self.root * VOLTS_PER_SEMITONE
+        
+        self.output_voltage = volts
+        self.current_note = semitone
+       
+    def read_quantize_output(self):
+        """Read the input signal, quantize it, set outputs 1-5 accordingly
+
+        Called by the main loop in continuous mode or the rising clock handler
+        in triggered mode
+        """
+        self.input_voltage = ain.read_voltage(128)
+        self.quantize(self.input_voltage)
+        
+        cv1.voltage(self.output_voltage)
+        
+        for i in range(len(self.aux_outs)):
+            self.aux_outs[i].voltage(self.output_voltage + self.intervals[i] * VOLTS_PER_SEMITONE)
+    
+    def main(self):
+        """The main loop; reads from ain, sets the output voltages
+        """
+        
+        while True:
+            # Update at 100Hz
+            CYCLE_RATE = 0.01
+            time.sleep(CYCLE_RATE)
+            
+            # Check if we've been idle for too long; if so, blank the screen
+            # to prevent burn-in
+            now = time.ticks_ms()
+            if time.ticks_diff(now, self.last_interaction_time) > SCREENSAVER_TIMEOUT_MS:
+                self.active_screen = self.screensaver
+            
+            # read the encoder value from knob 1 and apply it to the
+            # active menu items for the UI
+            self.highlight_note = round(knob_rescale(k1, 0, len(self.scale)-1))
+            self.menu_item = round(knob_rescale(k1, 0, len(self.menu.menu_items)-1))
+            
+            if self.mode == MODE_CONTINUOUS:
+                cv6.off()
+                last_output = self.output_voltage
+                self.read_quantize_output()
+                
+                if last_output != self.output_voltage:
```

Comment:
[discussion] I'm seeing some strange behavior, that I'm having trouble explaining. 

Here's my setup: I have the quantizer patched up in such a way that that I'm feeding it a constant voltage to ain that I can control via a knob on another module. I'm sending cv1 to the pitch of an oscillator and am listing to the output. I then tune the input voltage to a place where the quantizer has trouble choosing between two notes. It's easier to do this if you turn off a bunch of notes in the scale and put the voltage in the large empty space you created.

The fact that the note jumps around is not the strange part. It is known that the voltages read by ain on this hardware can be noisy, and the code does nothing to mitigate it. We can iterate on this later to attempt to tame it.

The strangeness is that the note jumps around at a different rate depending on which screen I'm on. I've been able to 'tune' it such that the note is stable when viewing the piano, but erratic when viewing the menu, and vice versa.

I've looked through the code and don't see anything in the menu that would directly affect the quantization. Since you are considerably more familiar with this code, I thought that I would point this out to you in case you had any ideas. 

My only idea is that the drawing of the piano takes more time than the menu, causing a slightly different rate of quantization depending on what screen is being viewed. And this rate somehow interacts with something else on the pico, such as the rate of whatever is causing the ain noise.

I don't think that this necessarily needs to be fixed before merge, but it is strange enough that I thought that it might point towards a subtle bug that I just can't see yet.

Reply:
One possible solution here is to reduce the frequency of draw calls by introducing a "dirty" flag to indicate if the UI has changed, and only draw the screen if "dirty" is `True`. You can see as similar implementation in the Polyrhythmic Seq script where I prevent unnecessary saving if state has not changed or minimum duration has not yet passed: https://github.com/Allen-Synthesis/EuroPi/blob/759478558332e47552f34cd2b4e5cd5d16df1fb7/software/contrib/polyrhythmic_sequencer.py#L236

Reply:
Yeah, the noise-mitigation is an issue. The only thing I've really done to address it was to increase the samples when reading `ain` in an attempt to smooth it slightly. But if your voltage is on the boundary between rounding up vs rounding down then you will get some jitter with the current implementation.

I wouldn't be surprised if the keyboard takes significantly longer to draw; it's a bitmap layer with text super-imposed over it with a couple of loops, as opposed to _just_ some text.  I suspect the bit-blitting to get the keyboard background layer uses most of the difference in CPU cycles.

De-noising is definitely something that's in the queue for the next iteration. Or if someone with more experience doing noise-reduction on a low-power system wants to tackle it I'm happy to accept MRs back to my source branch too.